src/main/java/graphql/schema/transform/FieldVisibilitySchemaTransformation.java

```diff
@@ -0,0 +1,127 @@
+                new TypeVisibilityVisitor(observedTypes, removedTypes));
+
+        afterTransformation();
+
+        return finalSchema;
+    }
+
+    private static class FieldVisibilityVisitor extends GraphQLTypeVisitorStub {
+
+        private final VisibleFieldPredicate visibilityPredicate;
+        private final Set<GraphQLType> removedTypes;
+        private final Set<GraphQLType> observedTypes;
+
+        private FieldVisibilityVisitor(VisibleFieldPredicate visibilityPredicate,
+                               Set<GraphQLType> removedTypes, Set<GraphQLType> observedTypes) {
+            this.visibilityPredicate = visibilityPredicate;
+            this.removedTypes = removedTypes;
+            this.observedTypes = observedTypes;
+        }
+
+        @Override
+        public TraversalControl visitGraphQLInterfaceType(GraphQLInterfaceType node,
+                                                          TraverserContext<GraphQLSchemaElement> context) {
+            observedTypes.add(node);
+
+            return TraversalControl.CONTINUE;
+        }
+
+        @Override
+        public TraversalControl visitGraphQLObjectType(GraphQLObjectType node,
+                                                       TraverserContext<GraphQLSchemaElement> context) {
+
+            if (context.getBreadcrumbs().stream()
+                    .anyMatch(crumb -> crumb.getLocation().getName().equalsIgnoreCase("addTypes"))) {
+                return TraversalControl.ABORT;
+            }
+
+            observedTypes.add(node);
+
+            return TraversalControl.CONTINUE;
+        }
+
+        @Override
+        public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition definition,
+                                                            TraverserContext<GraphQLSchemaElement> context) {
+            if (!visibilityPredicate.isVisible(definition)) {
+                deleteNode(context);
+                removedTypes.add(definition.getType());
+            }
+
+            return TraversalControl.CONTINUE;
+        }
+    }
+
+    private static class TypeVisibilityVisitor extends GraphQLTypeVisitorStub {
+
+        private final Set<GraphQLType> observedTypes;
+        private final Set<GraphQLType> removedTypes;
+
+        private TypeVisibilityVisitor(Set<GraphQLType> observedTypes, Set<GraphQLType> removedTypes) {
+            this.observedTypes = observedTypes;
+            this.removedTypes = removedTypes;
+        }
+
+        @Override
+        public TraversalControl visitGraphQLObjectType(GraphQLObjectType node,
+                                                       TraverserContext<GraphQLSchemaElement> context) {
+            if (!observedTypes.contains(node) &&
+                    node.getInterfaces().stream().anyMatch(removedTypes::contains) &&
+                    !ScalarInfo.isStandardScalar(node.getName()) &&
+                    !node.getName().equalsIgnoreCase("Query")) {
```

Comment:
schemas don't have to have a type named Query. The "query" type can be any type defined as the query type.

Reply:
we should also add tests for these kinds of schema. See https://graphql.github.io/graphql-spec/draft/#sec-Root-Operation-Types for some background.

Reply:
> The "query" type can be any type
How do I determine whether the node is the query type then?

Reply:
You need to save the original `schema.getQueryType()` and do either a object map or better yet name match (since names of types are unique in graphql)

Reply:
Updated.

Reply:
And test added.